### PR TITLE
[SPARK-50034][CORE] Fix Misreporting of Fatal Errors as Uncaught Exceptions in `SparkUncaughtExceptionHandler`

### PR DESCRIPTION
**Pull request by mingkangli-db**
#### Summary

This PR addresses the misreporting of fatal errors (such as `OutOfMemoryError`) as uncaught exceptions in `SparkUncaughtExceptionHandler`.

#### Context

In `Executor.scala`, an exception is considered fatal if any exception in the chain (or its causes) is fatal. This behavior is controlled by the `spark.executor.killOnFatalError.depth` config, which limits the depth to which the chain is inspected. If a fatal error is found, `SparkUncaughtExceptionHandler` is called.

Currently, `SparkUncaughtExceptionHandler` only considers the top-level exception when reporting the exit code, rather than traversing the full exception chain to identify the true fatal cause. As a result, some fatal errors, such as `OutOfMemoryError`, are mistakenly reported as uncaught exceptions.

For instance, if we have an OOM exception with the following structure:

```
RuntimeException
 - Caused by: RuntimeException
 - Caused by: java.lang.OutOfMemory
```

SparkUncaughtExceptionHandler would quit the executor with error code SparkExitCode.UNCAUGHT_EXCEPTION, when the true cause is an OOM error.

#### Fix

This PR modifies `SparkUncaughtExceptionHandler` to:
- Inspect the exception chain (up to the configured depth).
- Ensure that the actual fatal error is correctly identified and reflected in the exit code.

### Why are the changes needed?

This is a bug because some nonfatal exceptions are mistakenly reported as an uncaught exception when the true cause is some other fatal error down the exception chain.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit tests are added to `SparkUncaughtExceptionHandlerSuite.scala`.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala [hunk 1]
```diff
@@ -36,6 +38,20 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
...
 
+  // The maximum depth to search in the exception cause chain for a fatal error,
+  // as defined by killOnFatalErrorDepth in Executor.scala.
+  //
+  // SPARK-50034: When this handler is called, there is a fatal error in the cause chain within
+  // the specified depth. We should identify that fatal error and exit with the
+  // correct exit code.
+  private val killOnFatalErrorDepth: Int = {
+    // At this point SparkEnv might be None
+    Option(SparkEnv.get) match {
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:49]: nit: Option(SparkEnv.get).map(...).getOrElse(5)

File: core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala [hunk 2]
```diff
@@ -74,6 +80,34 @@ object ThrowableTypes extends Enumeration {
...
+
+  val NestedSparkFatalException = ThrowableTypesVal(
+    "DeepNestedMixedFatalErrors",
+    new RuntimeException("Nonfatal Level 1",
+      new RuntimeException("Nonfatal Level 2",
+        new SparkFatalException(new OutOfMemoryError())))
+  )
+
+  // Nested exception with non-fatal errors only
+  val NonFatalNestedErrors = ThrowableTypesVal(
```
Comment (jiangxb1987) [inline core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala:102]: Should we also test the OOME under deeply nested exceptions, that exceeded the killOnFatalErrorDepth?

**Review by jiangxb1987 — APPROVED** [review]
LGTM only a few nits

File: core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala (hunk 2 above)
Comment (mingkangli-db) [reply core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala:104]: Good point. Added a test case to cover this.

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala [hunk 3]
```diff
@@ -36,6 +38,17 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
...
 
+  // The maximum depth to search in the exception cause chain for a fatal error,
+  // as defined by killOnFatalErrorDepth in Executor.scala.
+  //
+  // SPARK-50034: When this handler is called, there is a fatal error in the cause chain within
+  // the specified depth. We should identify that fatal error and exit with the
+  // correct exit code.
+  private val killOnFatalErrorDepth: Int =
+    // At this point SparkEnv might be None
+    Option(SparkEnv.get).map(_.conf.get(EXECUTOR_KILL_ON_FATAL_ERROR_DEPTH)).getOrElse(5)
```
Comment (zeal-thinker) [inline core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:49]: Why do we have EXECUTOR in EXECUTOR_KILL_ON_FATAL_ERROR_DEPTH name? We are sometimes killing master\worker depending upon the value of EXECUTOR_KILL_ON_FATAL_ERROR_DEPTH as master and worker use SparkUncaughtExceptionHandler.

Comment (zeal-thinker) [inline core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:49]: How did you decide on the default value to be 5?

Comment (JiexingLi) [reply core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:49]: Yes, let's remove the executor from the name? SparkUncaughtExceptionHandler is also used by master.

Comment (mingkangli-db) [reply core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:49]: Sounds good, I'll remove "EXECUTOR" from the name

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala [hunk 4]
```diff
@@ -36,6 +38,17 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
...
   }
 
+  // The maximum depth to search in the exception cause chain for a fatal error,
+  // as defined by killOnFatalErrorDepth in Executor.scala.
+  //
+  // SPARK-50034: When this handler is called, there is a fatal error in the cause chain within
+  // the specified depth. We should identify that fatal error and exit with the
+  // correct exit code.
+  private val killOnFatalErrorDepth: Int =
+    // At this point SparkEnv might be None
```
[the remark this answers is not in the document]
Comment (mingkangli-db) [reply core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:48]: When registering the uncaught exception handler on the `Master` ([code](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L1383-1384)) or `Worker` ([code](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala#L981-L982)), `SparkEnv` is not yet initialized, while it is initialized on the Executor, so we have account for both cases.

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala (hunk 3 above)
Comment (zeal-thinker) [reply core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala]: Okay! Thanks

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 5]
```diff
@@ -177,7 +177,7 @@ private[spark] class Executor(
   // Whether to monitor killed / interrupted tasks
   private val taskReaperEnabled = conf.get(TASK_REAPER_ENABLED)
 
-  private val killOnFatalErrorDepth = conf.get(EXECUTOR_KILL_ON_FATAL_ERROR_DEPTH)
+  private val killOnFatalErrorDepth = conf.get(KILL_ON_FATAL_DEPTH)
```
Comment (zeal-thinker) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala:180]: Minor: Why did you remove ERROR from KILL_ON_FATAL_DEPTH name?

Comment (zeal-thinker) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala:180]: Minor: Why did you remove ERROR from KILL_ON_FATAL_DEPTH name?

Comment (zeal-thinker) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala:180]: Minor: Why did you remove ERROR from KILL_ON_FATAL_DEPTH?

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala [hunk 6]
```diff
@@ -36,6 +38,17 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
     val _ = SparkExitCode.OOM
   }
 
+  // The maximum depth to search in the exception cause chain for a fatal error,
+  // as defined by killOnFatalErrorDepth in Executor.scala.
+  //
```
Comment (zeal-thinker) [inline core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:43]: Minor: Want to remove Executor.scala as KILL_ON_FATAL_DEPTH is defined in package.scala?

File: core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala [hunk 7]
```diff
@@ -74,6 +82,46 @@ object ThrowableTypes extends Enumeration {
...
+  val NestedOOMError = ThrowableTypesVal(
+    "NestedFatalError",
+    new RuntimeException("Nonfatal Level 1",
+      new RuntimeException("Nonfatal Level 2",
+        new RuntimeException("Nonfatal Level 3",
+          new OutOfMemoryError())))
+  )
+
+  val NestedSparkFatalException = ThrowableTypesVal(
+    "NestedMixedFatalErrors",
```
Comment (zeal-thinker) [inline core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala:97]: What is mixed fatal error?

Comment (mingkangli-db) [reply core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala]: Changed to `NestedSparkFatalException`. I changed the variable name but forgot to change the name inside `ThrowableTypesVal`. Fixed.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 5 above)
Comment (mingkangli-db) [reply core/src/main/scala/org/apache/spark/executor/Executor.scala]: Good catch. Added back ERROR.

**Comment by mingkangli-db** [comment]
@zeal-thinker @JiexingLi Thank you for the feedback! I've addressed all the comments and I think it's ready for another round of reviews :)

**Review by zeal-thinker — APPROVED** [review]
LGTM

**Comment by zeal-thinker** [comment]
> @zeal-thinker @JiexingLi Thank you for the feedback! I've addressed all the comments and I think it's ready for another round of reviews :)

Thank you for addressing all comments!

**Review by JiexingLi — APPROVED** [review]

**Comment by JiexingLi** [comment]
jenkins merge

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala [hunk 8]
```diff
@@ -50,19 +63,30 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
...
+              System.exit(ExecutorExitCode.KILLED_BY_TASK_REAPER)
+            // No match, continue traversing the cause chain
+            case _ =>
+          }
+          // Move to the next cause in the chain
+          currentException = currentException.getCause
+          depth += 1
+        }
+
+        if (exitOnUncaughtException) {
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:88]: Let's keep the comment:
```suggestion
        // SPARK-30310: Don't System.exit() when exitOnUncaughtException is false
        if (exitOnUncaughtException) {
```

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala [hunk 9]
```diff
@@ -50,19 +63,31 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
...
-            // SPARK-24294: This is defensive code, in case that SparkFatalException is
-            // misused and uncaught.
-            System.exit(SparkExitCode.OOM)
-          case _: KilledByTaskReaperException if exitOnUncaughtException =>
-            System.exit(ExecutorExitCode.KILLED_BY_TASK_REAPER)
-          case _ if exitOnUncaughtException =>
-            System.exit(SparkExitCode.UNCAUGHT_EXCEPTION)
-          case _ =>
-            // SPARK-30310: Don't System.exit() when exitOnUncaughtException is false
+        // Traverse the causes up to killOnFatalErrorDepth layers
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:66]: > In Executor.scala, an exception is considered fatal if any exception in the chain (or its causes) is fatal. 

Can we reuse the code there? Like adding a function `def findRootCauseError`

Comment (mingkangli-db) [reply core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:66]: Oops, I think we should address this separately in a follow-up ticket because I just realized another issue with uncaught exception handling. In `Executor.scala`, the class `KilledByTaskReaperException(message: String) extends SparkException(message)`, added in [SPARK-48541](https://github.com/apache/spark/pull/46883), is not considered a fatal error when checked through `Executor.isFatalError()`:

For example, running the following code returns `false`:

```scala
// Create an instance of the KilledByTaskReaperException
val exception = new KilledByTaskReaperException("Task was killed by the task reaper")
// Result is false!
val result = Executor.isFatalError(exception, depthToCheck = 5)
```

However, in `SparkUncaughtExceptionHandler`, it is assumed to be a fatal error (see the code above):

```scala
case _: KilledByTaskReaperException if exitOnUncaughtException =>
    System.exit(ExecutorExitCode.KILLED_BY_TASK_REAPER)
```

Additionally, we seem to assume it should be fatal error here in `Executor.scala` as well, from the comments: [code link](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L1043-L1048). Seems like we expect the exception to bubble up to the `uncaughtExceptionHandler` here: [code link](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L861-L862), but it won't because `Executor.isFatalError()` returns `false`!

This inconsistency arises because `Executor.isFatalError` doesn't treat `KilledByTaskReaperException` as fatal (when it should!), but we assumed in `Executor.scala` and in `SparkUncaughtExceptionHandler` that it is fatal. To reuse the logic, we need to fix `Executor.isFatalError` to recognize `KilledByTaskReaperException` as a fatal error first. 

What do you think? @cloud-fan

Cc:  @bozhang2820 (who added `KilledByTaskReaperException`)

Comment (cloud-fan) [reply core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:66]: So the code in `Executor.scala` is
```
  def isFatalError(t: Throwable, depthToCheck: Int): Boolean = {
    if (depthToCheck <= 0) {
      false
    } else {
      t match {
        case _: SparkOutOfMemoryError => false
        case e if Utils.isFatalError(e) => true
        case e if e.getCause != null => isFatalError(e.getCause, depthToCheck - 1)
        case _ => false
      }
    }
  }
```

It seems hard to reuse, let's leave it for now

**Comment by cloud-fan** [comment]
thanks, merging to master/3.5!

File: core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala [hunk 10]
```diff
@@ -36,6 +38,17 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
     val _ = SparkExitCode.OOM
   }
 
+  // The maximum depth to search in the exception cause chain for a fatal error,
+  // as defined by KILL_ON_FATAL_ERROR_DEPTH.
+  //
+  // SPARK-50034: When this handler is called, there is a fatal error in the cause chain within
+  // the specified depth. We should identify that fatal error and exit with the
+  // correct exit code.
+  private val killOnFatalErrorDepth: Int =
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:47]: nit: Could have been a lazy val ... we dont need it as part of initialization.
